server/src/main/java/org/cloudfoundry/identity/uaa/oauth/KeyInfo.java

```diff
@@ -19,298 +22,169 @@
-                RSAPublicKeySpec pubSpec = new RSAPublicKeySpec(key.getModulus(), key.getPublicExponent());
-                RSAPrivateCrtKeySpec privSpec = new RSAPrivateCrtKeySpec(
-                  key.getModulus(),
-                  key.getPublicExponent(),
-                  key.getPrivateExponent(),
-                  key.getPrime1(),
-                  key.getPrime2(),
-                  key.getExponent1(),
-                  key.getExponent2(),
-                  key.getCoefficient()
-                );
-                publicKey = fact.generatePublic(pubSpec);
-                privateKey = fact.generatePrivate(privSpec);
-            } else if ("PUBLIC KEY".equals(type)) {
-                KeySpec keySpec = new X509EncodedKeySpec(content);
-                publicKey = fact.generatePublic(keySpec);
-            } else if ("RSA PUBLIC KEY".equals(type)) {
-                ASN1Sequence seq = ASN1Sequence.getInstance(content);
-                org.bouncycastle.asn1.pkcs.RSAPublicKey key = org.bouncycastle.asn1.pkcs.RSAPublicKey.getInstance(seq);
-                RSAPublicKeySpec pubSpec = new RSAPublicKeySpec(key.getModulus(), key.getPublicExponent());
-                publicKey = fact.generatePublic(pubSpec);
-            } else {
-                throw new IllegalArgumentException(type + " is not a supported format");
             }
-
-            return new KeyPair(publicKey, privateKey);
-        } catch (InvalidKeySpecException e) {
-            throw new RuntimeException(e);
-        } catch (NoSuchAlgorithmException e) {
-            throw new IllegalStateException(e);
+            if (type == RSA) {
+                RSAPublicKey rsaKey;
+                try {
+                    result.put(JsonWebKey.PUBLIC_KEY_VALUE, this.verifierKey);
+                    rsaKey = jwk.toRSAKey().toRSAPublicKey();
+                } catch (JOSEException e) {
+                    throw new IllegalArgumentException(e);
+                }
+                String n = Base64URL.encode(rsaKey.getModulus()).toString();
+                String e = Base64URL.encode(rsaKey.getPublicExponent()).toString();
+                result.put(JWKParameterNames.RSA_MODULUS, n);
+                result.put(JWKParameterNames.RSA_EXPONENT, e);
+            } else if (type == EC) {
+                result.putAll(jwk.toJSONObject());
+            }
+            return result;
+        } else {
+            Map<String, Object> result = new HashMap<>();
+            result.put(HeaderParameterNames.ALGORITHM, this.algorithm());
+            result.put(JsonWebKey.PUBLIC_KEY_VALUE, this.verifierKey);
+            //new values per OpenID and JWK spec
+            result.put(JWKParameterNames.PUBLIC_KEY_USE, JsonWebKey.KeyUse.sig.name());
+            result.put(HeaderParameterNames.KEY_ID, this.keyId);
+            result.put(JWKParameterNames.KEY_TYPE, type.name());
```

Comment:
Would it make sense to move these lines (except for the one setting `PUBLIC_KEY_VALUE`) above the `if`-statement? Then we could also get rid of them in the first lines of the other branch.